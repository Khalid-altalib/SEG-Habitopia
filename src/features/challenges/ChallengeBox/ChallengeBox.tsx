import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { Box, Heading, HStack } from "native-base";
import React from "react";
import { TouchableOpacity, Image } from "react-native";
import { Challenge, RootParams } from "../../../../types";
import { challengeMappings } from "../challengeMappings";
import ChallengeOnGoingText from "../ChallengeOnGoingText/ChallengeOnGoingText";

type Props = {
  challenge: Challenge;
};

const ChallengeBox = (props: Props) => {
  const { challenge } = props;
  const { name, active } = challenge;

  const navigation = useNavigation<NativeStackNavigationProp<RootParams>>();

  const { color: cardColor, image: cardImage } =
    challengeMappings[name] || challengeMappings["fallback"];

  const handlePress = () =>
    navigation.navigate("ChallengePrompt", {
      challenge: challenge,
    });

  return (
    <TouchableOpacity onPress={handlePress}>
      <Box
        backgroundColor={cardColor}
        borderRadius={"md"}
        shadow={3}
        padding={3}
        height={100}
      >
        <HStack justifyContent="space-between" height="100%">
          <Box justifyContent="space-between">
            <Heading>{name}</Heading>
            <ChallengeOnGoingText onGoing={active} />
          </Box>
          <Box borderRadius={3}>
            <Image
              source={{ uri: cardImage }}
              borderRadius={6}
              style={{ height: "100%", width: 80 }}
            />
          </Box>
        </HStack>
      </Box>
    </TouchableOpacity>
  );
};

export default ChallengeBox;
