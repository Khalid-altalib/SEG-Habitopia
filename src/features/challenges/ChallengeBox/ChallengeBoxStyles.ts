import { background, border } from "native-base/lib/typescript/theme/styled-system";
import { StyleSheet } from "react-native";

const challengeBoxStyles = StyleSheet.create({
unactiveChallengeBox: {
    flex: 1,
    width: '100%',
    overflow: 'hidden',
    margin: 5,
    borderRadius: 20,
    shadowColor: '#000',
    justifyContent: 'center',
    shadowOffset: {
        width: 0,
        height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
    borderWidth: 4,
    minHeight:80,
},
activeChallengeBox:{
    flex: 1,
    width: '100%',
    overflow: 'hidden',
    margin: 5,
    borderRadius: 20,
    shadowColor: '#000',
    justifyContent: 'center',
    shadowOffset: {
        width: 0,
        height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
    borderColor: 'red',
    borderWidth: 4,
    minHeight:80,
},
challengeText: {
    fontSize: 25,
},
onGoingText: {
    color: 'red'
},
challengeImage: {
    height: 80,
    width: 80,
    marginRight: 0,
    marginLeft: 'auto',
}
});

export default challengeBoxStyles;