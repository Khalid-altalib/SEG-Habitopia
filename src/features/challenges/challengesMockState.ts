import { ChallengesState } from "./challengesSlice";

// Mocks the challenge state for testing purposes
const challengesMockState: ChallengesState = {
    challenges: [
        {
            id: '1',
            name: 'Fitness',
            description: 'Be active',
            active: true

        },
        {
            id: '1',
            name: 'Study',
            description: 'Study hard',
            active: false
        }
    ],
    fetchChallenges: {
        loading: false,
        error: ""
    },
    joinChallenge: {
        loading: false,
        error: ""
    }
};

export default challengesMockState;