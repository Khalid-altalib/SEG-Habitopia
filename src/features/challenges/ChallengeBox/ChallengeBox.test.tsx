// Import required modules
import { fireEvent, render } from "@testing-library/react-native";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import TestingWrapperNavigation from "@app/testingWrapperWithNavigation";
import challengesMockState from "../challengesMockState";
import ChallengeBox from "./ChallengeBox";
import { extendTheme, NativeBaseProvider } from "native-base";

// Start test suite for ChallengeBox component
describe("ChallengeBox", () => {
  // Set up mock state and store for testing
  const mockState = {
    challenges: challengesMockState,
  };
  const mockStore = configureStore([thunk])(mockState);

  // Set up props for active and inactive challenge boxes
  const activeBoxProps = {
    challenge: mockState.challenges.challenges[0],
  };
  const unactiveBoxProps = {
    challenge: mockState.challenges.challenges[1],
  };

  // Test that the active challenge box renders correctly
  it("renders the active challenge box correctly", async () => {
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <ChallengeBox {...activeBoxProps} />
      </TestingWrapperNavigation>
    );

    expect(component).toMatchSnapshot();
  });

  // Test that the inactive challenge box renders correctly
  it("renders the inactive challenge box correctly", async () => {
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <ChallengeBox {...unactiveBoxProps} />
      </TestingWrapperNavigation>
    );

    expect(component).toMatchSnapshot();
  });

  // Test that the challenge box can be found by test ID
  it("finds the challenge box", async () => {
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <ChallengeBox {...activeBoxProps} />
      </TestingWrapperNavigation>
    );

    const box = component.getByTestId("challengeBox");

    expect(box).toBeDefined();
  });

  // Test that clicking on the challenge box opens the challenge prompt
  it("clicking on the challenge box brings up the challenge prompt", async () => {
    // Set up theme and window metrics for NativeBaseProvider
    const theme = extendTheme({});
    const inset = {
      frame: { x: 0, y: 0, width: 0, height: 0 },
      insets: { top: 0, left: 0, right: 0, bottom: 0 },
    };

    // Render the ChallengeBox component within the TestingWrapperNavigation and NativeBaseProvider
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <NativeBaseProvider theme={theme} initialWindowMetrics={inset}>
          <ChallengeBox {...activeBoxProps} />
        </NativeBaseProvider>
      </TestingWrapperNavigation>
    );

    // Get the challenge box element and simulate a press event
    const box = component.getByTestId("challengeBox");
    fireEvent(box, "press");

    // Get the challenge prompt element and test that it is defined
    const newScreen = component.getByTestId("challengePrompt");
    expect(newScreen).toBeDefined();
  });
});
