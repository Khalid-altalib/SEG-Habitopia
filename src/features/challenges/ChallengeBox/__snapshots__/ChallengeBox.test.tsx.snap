// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeBox renders the active challenge box correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="challengeBox"
            >
              <View
                dataSet={{}}
                style={
                  {
                    "backgroundColor": "#fbbf24",
                    "borderRadius": 8,
                    "elevation": 4,
                    "height": 100,
                    "overflow": "hidden",
                    "shadowColor": "#000000",
                    "shadowOffset": {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.23,
                    "shadowRadius": 2.62,
                  }
                }
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "flexDirection": "row",
                      "height": "100%",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <View
                    dataSet={{}}
                    style={
                      {
                        "justifyContent": "space-between",
                        "paddingBottom": 12,
                        "paddingLeft": 12,
                        "paddingRight": 12,
                        "paddingTop": 12,
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 23,
                          },
                          undefined,
                        ]
                      }
                    >
                      Fitness
                    </Text>
                    <View
                      dataSet={{}}
                      style={
                        {
                          "backgroundColor": "#ef4444",
                          "borderRadius": 4,
                          "paddingBottom": 4,
                          "paddingLeft": 8,
                          "paddingRight": 8,
                          "paddingTop": 4,
                        }
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "white",
                            },
                            {
                              "fontFamily": undefined,
                              "fontSize": 19,
                            },
                            undefined,
                          ]
                        }
                      >
                        On-going
                      </Text>
                    </View>
                  </View>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <View
                    dataSet={{}}
                    style={
                      [
                        {},
                        {
                          "aspectRatio": 1,
                          "style": undefined,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        {
                          "uri": "https://picsum.photos/200",
                        }
                      }
                      style={
                        {
                          "bottom": 0,
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;

exports[`ChallengeBox renders the unactive challenge box correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="challengeBox"
            >
              <View
                dataSet={{}}
                style={
                  {
                    "backgroundColor": "#c084fc",
                    "borderRadius": 8,
                    "elevation": 4,
                    "height": 100,
                    "overflow": "hidden",
                    "shadowColor": "#000000",
                    "shadowOffset": {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.23,
                    "shadowRadius": 2.62,
                  }
                }
              >
                <View
                  dataSet={{}}
                  style={
                    {
                      "flexDirection": "row",
                      "height": "100%",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <View
                    dataSet={{}}
                    style={
                      {
                        "justifyContent": "space-between",
                        "paddingBottom": 12,
                        "paddingLeft": 12,
                        "paddingRight": 12,
                        "paddingTop": 12,
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 23,
                          },
                          undefined,
                        ]
                      }
                    >
                      Study
                    </Text>
                  </View>
                  <View
                    dataSet={{}}
                    style={{}}
                  />
                  <View
                    dataSet={{}}
                    style={
                      [
                        {},
                        {
                          "aspectRatio": 1,
                          "style": undefined,
                        },
                      ]
                    }
                  >
                    <Image
                      source={
                        {
                          "uri": "https://picsum.photos/200",
                        }
                      }
                      style={
                        {
                          "bottom": 0,
                          "left": 0,
                          "position": "absolute",
                          "right": 0,
                          "top": 0,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;
